app_confetti/config.py

```diff
@@ -1,21 +1,26 @@
+import dataclasses
+
+from app_confetti import util
+
+
+@dataclasses.dataclass(frozen=True)
 class BaseConfig:
     """Configuration class.
 
-    The `environ_config` library allows for its attributes to be retrieved from
-    environment variables, the prefix for this class and the attribute name
-    defines the variable name to be used, e.g. PREFIX_SENTRY_DSN for SENTRY_DSN.
+    The `env` util function allows for attributes to be retrieved from environment variables.
 
-    This class is generally used in settings.__init__.py to instantiate an object from
-    the environment.
-
-    Inheriting classes must implement:
-        cls.LOGGING_LEVEL
-        cls.SENTRY_DSN
-        cls.ENV
+    Inheriting classes can override:
+        logging_level: str
+        sentry_dsn: str
+        env: str
     """
 
+    logging_level: str = util.env("LOGGING_LEVEL:INFO")
+    sentry_dsn: int = util.env("SENTRY_DSN:None", util.str_to_literal)
```

Comment:
It's probably better to build a more explicit function that checks if the value is some sort of special token to denote "None" such as `__NONE`, otherwise returning the string itself. We might get lots of implicit behaviour using `ast.parse_literal` plus there's no real reason to use `ast` unless you're building one. You'll also want to state the type as `typing.Optional[T]` given it may not be defined.

Reply:
parse_literal is mostly there for pulling dictionaries and lists out of strings without having to resort to json. As the config you are pulling in is controlled by you and not arbitrary data outside the users control, str_to_literal shouldn't give any unintended functionality unless ENV vars are set up wrong.

Reply:
Nullable options should be handled differently though, agreed.

Reply:
If you're pulling lists/dicts out of strings, why _not_ use json?

```python
In [1]: import json

In [2]: json.loads("[1, 2, 3]")
Out[2]: [1, 2, 3]

In [3]: json.loads('{"foo": "bar"}')
Out[3]: {'foo': 'bar'}
```

These are all valid.

Reply:
it's not valid for `'{"foo": True}'` etc, it depends on what you are storing the config in, but that plays to your point, the function provided should be specific to the case being handled.